### DISCUSSION TITLE
OpenVINO™ ARM CPU plugin

### DISCUSSION THREAD

**Discussion by atlury** [blakeblackshear/frigate#5534] [Ideas]
Hello all,

Great job getting being done on this project. Much appreciated. I have a suggestion. 

OpenVINO has an ARM CPU plugin that can do inference on ARM CPUs, using OpenVINO API. The plugin uses the ARM Compute Library (https://github.com/ARM-software/ComputeLibrary) as a backend. This is from 

https://github.com/openvinotoolkit/openvino_contrib/blob/master/modules/arm_plugin/README.md

and here

https://docs.openvino.ai/latest/openvino_docs_OV_UG_supported_plugins_ARM_CPU.html

and lastly this

https://opencv.org/how-to-run-yolov4-using-openvino-and-opencv-on-arm/

Since, we already support openvino x86, can we also think of testing the arm version on Raspberry Pi 4? Any takers?

Thanking you

**Comment by NateMeyer** [comment]
The ARM plugin is included in builds for both arm32 and arm64.  Feel free to test it out!

**Comment by atlury** [comment]
Great but the beta documentation doesnt mention much on this, only says (from here https://deploy-preview-4055--frigate-docs.netlify.app/frigate/hardware) 

> The OpenVINO detector type is able to run on:
> 
> 6th Gen Intel Platforms and newer that have an iGPU
> x86 & Arm32/64 hosts with VPU Hardware (ex: Intel NCS2)

Any specific instructions for docker-compose and config file "other" than the regular ones below for openvino arm?

> version: "3.9"
> services:
>   frigate:
>     container_name: frigate
>     privileged: true # this may not be necessary for all setups
>     restart: unless-stopped
>     image: ghcr.io/blakeblackshear/frigate:0.12.0-beta8
>     shm_size: "64mb" # update for your cameras based on calculation above
>     devices:
>       - /dev/bus/usb:/dev/bus/usb # passes the USB Coral, needs to be modified for other versions
>       - /dev/apex_0:/dev/apex_0 # passes a PCIe Coral, follow driver instructions here https://coral.ai/docs/m2/get-started/#2a-on-linux
>       - /dev/dri/renderD128 # for intel hwaccel, needs to be updated for your hardware
>     volumes:
>       - /etc/localtime:/etc/localtime:ro
>       - /path/to/your/config.yml:/config/config.yml
>       - /path/to/your/storage:/media/frigate
>       - type: tmpfs # Optional: 1GB of memory, reduces SSD/SD Card wear
>         target: /tmp/cache
>         tmpfs:
>           size: 1000000000
>     ports:
>       - "5000:5000"
>       - "8554:8554" # RTSP feeds
>       - "8555:8555/tcp" # WebRTC over tcp
>       - "8555:8555/udp" # WebRTC over udp
>     environment:
>       FRIGATE_RTSP_PASSWORD: "password

and the config for detector below

> detectors:
>   ov:
>     type: openvino
>     device: AUTO
>     model:
>       path: /openvino-model/ssdlite_mobilenet_v2.xml
> 
> model:
>   width: 300
>   height: 300
>   input_tensor: nhwc
>   input_pixel_format: bgr
>   labelmap_path: /openvino-model/coco_91cl_bkgr.txt

**Reply by NateMeyer** [reply]
I think that might work. Best I can tell, the ARM plugin would be used as a CPU device, so you could also try setting that instead of AUTO.  If this is something that we can test and works well, we can add documentation for it with any details.

**Comment by atlury** [comment]
Thanks I will check and report back in a few days (or probably weeks) using CPU device.

**Comment by atlury** [comment]
Just add to this thread in-case someone is looking for the same. Nate is building by including openvino with arm cpu plugin and which means default CPU target and detected ARM architecture must tell openvino's inference engine to use ARM CPU plugin for inference.

https://github.com/NateMeyer/openvino-wheels/blob/main/arm_cpu_plugin_build.sh#L177

The ARM CPU plugin uses ARM Compute Library as a backend which should accelerate the inference

https://github.com/openvinotoolkit/openvino_contrib/blob/master/modules/arm_plugin/README.md

More later.

**Reply by cdennis12280** [reply]
Any updates on this? Does openvino work on pi4b + frigate?

**Comment by Welsyntoffie** [comment]
`detectors: 
  ov:
    type: openvino
    device: CPU
    model:
      path: /openvino-model/ssdlite_mobilenet_v2.xml

model:
  width: 300
  height: 300
  input_tensor: nhwc
  input_pixel_format: bgr
  labelmap_path: /openvino-model/coco_91cl_bkgr.txt`

Adding this to my Pi3B works, but inference shot up to 1500ms. Before this I used CPU detector with 3 threads and enference was at about 400, using a USB3 stick as boot drive.
